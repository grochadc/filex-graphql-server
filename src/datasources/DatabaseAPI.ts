import { ApolloError } from "apollo-server";
import { DataSource } from "apollo-datasource";
import { Workshop, Teacher, TeacherOption } from "../generated/graphql";
import { ParameterizedQuery as PQ } from "pg-promise";

export interface OptionModel {
  id: number;
  day: string;
  time: string;
  teacher_name: string;
  url: string;
  workshop_name: string;
  workshop_description: string;
  workshop_levels: number[];
  teacher_id: number;
  workshop_id: number;
  reservations: number;
}

export type StudentModel = {
  id: number;
  codigo: string;
  nombre: string;
  apellido_paterno: string;
  apellido_materno: string;
  genero: string;
  carrera: string;
  ciclo: string;
  telefono: string;
  email: string;
  nivel: string;
  curso: string;
  externo: boolean;
  grupo: string;
};

type Option = Omit<
  OptionModel,
  "id" | "workshop_id" | "teacher_id" | "reservations"
> & {
  id: string;
  workshop_id: string;
  teacher_id: string;
  isTutorial: boolean;
  available: boolean;
};

export const SELECT_STUDENT = `SELECT * FROM student WHERE codigo=$1`;

export const SELECT_STUDENT_RESERVATION = `
SELECT
  r.*,
  o.day,
  o.time,
  o.url,
  o.teacher_id,
  t.name as teacher_name,
  o.workshop_id,
  w.name as workshop_name
FROM reservation r
INNER JOIN option o ON o.id=r.option_id
INNER JOIN teacher t ON t.id=o.teacher_id
INNER JOIN workshop w ON w.id=o.workshop_id
WHERE r.student_id=$1
`;

export const SELECT_NEW_RESERVATION = `
SELECT
  r.id,
  o.day,
  o.time,
  t.name as teacher_name,
  t.id as teacher_id,
  w.name as workshop_name,
  w.id as workshop_id,
  o.url
FROM reservation r
INNER JOIN option o ON r.option_id=o.id
INNER JOIN teacher t ON o.teacher_id=t.id
INNER JOIN workshop w ON o.workshop_id=w.id
WHERE r.id=$1
`;

export const SELECT_OPTIONS = `
SELECT
  Option.id,
  Workshop.name as workshop_name,
  Option.day,
  Option.time,
  Teacher.name as teacher_name,
  Option.url,
  Workshop.description as workshop_description,
  Workshop.levels as workshop_levels,
  Teacher.id as teacher_id,
  Workshop.id as workshop_id,
  COUNT(Reservation.id) as reservations
FROM Option
INNER JOIN Teacher ON Option.teacher_id=Teacher.id
INNER JOIN Workshop on Option.workshop_id=Workshop.id
LEFT JOIN Reservation ON Option.id=Reservation.option_id
GROUP BY Option.id, Workshop.id, Teacher.id
ORDER BY Option.id;
`;

export const SELECT_TEACHER_INFO = `
SELECT
  id,
  name,
  json_agg(
        (
        SELECT
          json_build_object('option_id',r.option_id, 'students', json_agg(s))
        FROM reservation r
        INNER JOIN student s ON s.id=r.student_id
        WHERE r.option_id=any(teacher.option_ids)
        GROUP BY r.option_id
        )
  ) as options FROM teacher Where id=$1 GROUP BY teacher.id;
`;

export type TeacherReservation = {
  reservation_id: number;
  codigo: string;
  nombre: string;
  apellido_materno: string;
  apellido_paterno: string;
  teacher_id: number;
  teacher: string;
  workshop: string;
  time: string;
  day: string;
  option_id: string;
};
export const SELECT_TEACHER_RESERVATIONS = `
SELECT
  t.id as teacher_id,
  r.id as reservation_id,
  codigo,
  nombre,
  apellido_paterno,
  apellido_materno,
  t.id as teacher_id,
  t.name as teacher,
  w.name as workshop,
  o.time,
  o.day,
  o.url,
  o.id as option_id
FROM teacher t
INNER JOIN option o ON o.teacher_id=t.id
INNER JOIN workshop w ON o.workshop_id=w.id
INNER JOIN reservation r ON r.option_id=o.id
INNER JOIN student s ON s.id=r.student_id
WHERE t.id=$1
`;

export const SELECT_SINGLE_OPTION = `
SELECT * FROM Option o WHERE o.id=$1;
`;

class DatabaseAPI extends DataSource {
  db: any;
  constructor(db: any) {
    super();
    console.log("construting databaseAPI");
    this.db = db;
  }
  _getStudentSql(query: string, values: any[]) {
    const paramQuery = new PQ({ text: query, values });
    return this.db.any(paramQuery);
  }
  async getStudent(codigo: string) {
    if (/^[1234567890]+$/.test(codigo)) {
      const result: StudentModel[] = await this._getStudentSql(SELECT_STUDENT, [
        codigo
      ]);
      if (result.length === 0)
        throw new ApolloError("404: Alumno inexistente", "404");
      return result[0];
    } else {
      throw new ApolloError("Input wasn't a number");
    }
  }
  async getOptionsByIDs(ids: string[]): Promise<Option> {
    const parsedIDs = ids.map(id => Number(id));
    const result = await this.db.any(
      new PQ({ text: SELECT_SINGLE_OPTION, values: [parsedIDs] })
    );
    return result[0];
  }
  async getAllOptions(max_reservations: number): Promise<Option[]> {
    const options: OptionModel[] = await this.db.any(SELECT_OPTIONS);
    const mapped = options.map(option => {
      return {
        ...option,
        available: Boolean(Number(option.reservations) < max_reservations),
        isTutorial: Boolean(option.workshop_id === 4),
        id: option.id.toString(),
        workshop_id: option.workshop_id.toString(),
        teacher_id: option.teacher_id.toString()
      };
    });
    return mapped;
  }

  async getAllWorkshops(max_reservations: number): Promise<Workshop[]> {
    const options = await this.getAllOptions(max_reservations);
    const workshops: Workshop[] = [];
    options.forEach(option => {
      const { workshop_description, ...simpleOption } = option;
      const foundWorkshopIndex = workshops.findIndex(
        workshop => workshop.id === option.workshop_id
      );
      if (foundWorkshopIndex === -1) {
        const new_workshop = {
          id: option.workshop_id,
          name: option.workshop_name,
          description: option.workshop_description,
          levels: option.workshop_levels.map(level => level.toString()),
          options: [simpleOption]
        };
        workshops.push(new_workshop);
      } else {
        workshops[foundWorkshopIndex].options.push(simpleOption);
      }
    });
    return workshops;
  }
  _getStudentReservationSql(query: string, values: any[]) {
    const paramQuery = new PQ({ text: query, values });
    return this.db.any(paramQuery);
  }
  getStudentReservation(id: number) {
    return this._getStudentReservationSql(SELECT_STUDENT_RESERVATION, [id]);
  }
  async _makeReservation(query: string, values: any[]) {
    const paramQuery = new PQ({ text: query, values });
    const result = await this.db.any(paramQuery);
    return result;
  }
  _newReservation(id: number) {
    return this.db.any(new PQ({ text: SELECT_NEW_RESERVATION, values: [id] }));
  }
  async makeReservation(
    student_id: number,
    option_id: number,
    tutorial_reason?: string
  ) {
    const INSERT_RESERVATION = `
      INSERT INTO reservation (student_id, option_id, tutorial_reason) VALUES ($1,$2,$3) RETURNING id;
    `;
    const normalizedTutorialReason =
      tutorial_reason === undefined ? null : tutorial_reason;
    const reservations = await this._makeReservation(INSERT_RESERVATION, [
      student_id,
      option_id,
      normalizedTutorialReason
    ]);
    const result = await this._newReservation(reservations[0].id);
    return result;
  }
  _resetReservations(query) {
    return this.db.any(query);
  }
  async resetReservations() {
    const DELETE_RESERVATIONS = `DELETE FROM reservation`;
    try {
      this._resetReservations(DELETE_RESERVATIONS);
    } catch (e) {
      throw new ApolloError(e);
    }
    return true;
  }
  _getTeacher(id: number) {
    const parsedQuery = new PQ({
      text: SIMPLE_TEACHER_INFO,
      values: [id]
    });
    return this.db.any(parsedQuery);
  }
  async getTeacher(id: string) {
    const parsedID = Number(id);
    const result: SimpleTeacherInfo[] = await this._getTeacher(parsedID);
    return {
      id: result[0].id.toString(),
      name: result[0].name,
      option_ids: result[0].option_ids.map(id => id.toString()),
      options: result.map(teacher => {
        return {
          id: teacher.option_id.toString(),
          day: teacher.day,
          time: teacher.time,
          teacher_name: teacher.name,
          teacher_id: teacher.id.toString(),
          workshop_name: teacher.workshop_name,
          workshop_id: teacher.workshop_id.toString(),
          url: teacher.url
        };
      })
    };
  }
  async getAllTeachers() {
    const result: SelectTeachers = await this.db.any(SELECT_TEACHERS);
    return result.map(teacher => {
      return {
        ...teacher,
        id: teacher.id.toString(),
        option_ids: teacher.option_ids.map(id => id.toString())
      };
    });
  }
  _getTeacherReservations(option_id: number) {
    return this.db.any(
      new PQ({ text: GET_TEACHER_RESERVATIONS, values: [option_id] })
    );
  }
  getTeacherReservations(option_id: string) {
    const parsedOptionID = Number(option_id);
    return this._getTeacherReservations(parsedOptionID);
  }
  _deleteOptionReservations(option_id: number) {
    return this.db.any(
      new PQ({ text: DELETE_TEACHER_RESERVATIONS, values: [option_id] })
    );
  }
  deleteOptionReservations(option_id: string) {
    const parsedID = Number(option_id);
    this._deleteOptionReservations(parsedID);
  }
}

export const GET_TEACHER_RESERVATIONS = `
SELECT
  r.id,
  w.id as workshop_id,
  w.name as workshop_name,
  o.id as option_id,
  s.codigo,
  s.nombre,
  s.apellido_paterno,
  s.apellido_materno,
  s.nivel,
  s.grupo,
  r.tutorial_reason
FROM reservation r
INNER JOIN option o ON r.option_id=o.id
INNER JOIN workshop w ON o.workshop_id=w.id
INNER JOIN student s ON s.id=r.student_id
WHERE o.id=$1
`;
export type SelectTeachers = {
  id: number;
  name: string;
  option_ids: number[];
}[];
export const SELECT_TEACHERS = `
SELECT * FROM Teacher;
`;

export type SimpleTeacherInfo = {
  id: number;
  name: string;
  workshop_name: string;
  workshop_id: number;
  option_id: number;
  day: string;
  time: string;
  url: string;
  option_ids: number[];
};
export const SIMPLE_TEACHER_INFO = `
SELECT t.id, t.name, w.name as workshop_name, w.id as workshop_id, o.id as option_id, o.day, o.time, o.url, t.option_ids
FROM teacher t
INNER JOIN option o ON o.teacher_id=t.id
INNER JOIN workshop w ON w.id=o.workshop_id
WHERE t.id=$1
`;

export const DELETE_TEACHER_RESERVATIONS = `
  DELETE FROM reservation WHERE option_id=$1
`;

export default DatabaseAPI;
