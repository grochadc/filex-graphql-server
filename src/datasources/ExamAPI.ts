import { DataSource } from "apollo-datasource";
import englishExam from "./data/exam_questions/en";
import frenchExam from "./data/exam_questions/fr";

class ExamAPI extends DataSource {
  getSection(course, level) {
    console.log("Getting questions for course ", course);
    if (course === "en") {
      return {
        totalSections: englishExam.sections.length,
        questions: englishExam.sections[level - 1].questions,
      };
    }
    if (course === "fr") {
      return {
        totalSections: frenchExam.sections.length,
        questions: frenchExam.sections[level - 1].questions,
      };
    } else {
      throw Error("That course doesn`t exist");
    }
  }
}
export { ExamAPI };
