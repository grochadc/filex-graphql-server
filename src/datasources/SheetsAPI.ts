import { DataSource } from "apollo-datasource";
import { google } from "googleapis";
require("dotenv").config();

function getJWT() {
  return new google.auth.JWT(
    process.env.GOOGLE_CLIENT_EMAIL,
    null,
    process.env.GOOGLE_PRIVATE_KEY.replace(/\\n/g, "\n"),
    ["https://www.googleapis.com/auth/spreadsheets"]
  );
}

class SheetsAPI extends DataSource {
  sheetID: any;
  api: any;
  ready: boolean;
  constructor(sheetID) {
    super();
    this.sheetID = sheetID;
    this.api = google.sheets({ version: "v4", auth: getJWT() });
  }

  async saveApplicant(applicant: Applicant) {
    const values = [
      [
        new Date().toUTCString(),
        applicant.id,
        applicant.codigo,
        applicant.nombre,
        applicant.apellido_paterno,
        applicant.apellido_materno,
        applicant.genero,
        applicant.ciclo,
        applicant.carrera,
        applicant.telefono,
        applicant.email,
        applicant.externo,
        applicant.reubicacion,
        applicant.curso,
        applicant.meetLink,
        applicant.nivel_escrito,
      ],
    ];
    this.api.spreadsheets.values.append({
      spreadsheetId: this.sheetID,
      range: "Today!A1",
      valueInputOption: "RAW",
      resource: {
        values,
      },
    });
  }
}

export { SheetsAPI };
