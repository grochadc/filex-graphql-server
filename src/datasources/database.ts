import pgPromise from "pg-promise";
import path from "path";
const { QueryFile } = pgPromise;
require("dotenv").config();

const initOptions = {
  connectionString: process.env.DATABASE_URL,
  ssl: { rejectUnauthorized: false }
};

console.log("Initializing database");
const pgp = pgPromise();
const db = pgp(initOptions);

export function sqlFile(file: string) {
  const fullPath = path.join(__dirname, file);
  const options = {
    minify: true
  };
  const qf = new QueryFile(fullPath, options);

  if (qf.error) {
    console.error(qf.error);
  }

  return qf;
}

export default db;
