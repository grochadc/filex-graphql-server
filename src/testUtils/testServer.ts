import {
  createTestClient,
  ApolloServerTestClient
} from "apollo-server-testing";
import { ApolloServer } from "apollo-server";

export default function testServer(
  dataSources?: () => any,
  context?: () => any
): ApolloServerTestClient {
  return createTestClient(
    new ApolloServer({
      modules: [
        require("../modules/placement_exam"),
        require("../modules/placement_settings"),
        require("../modules/registro"),
        require("../modules/workshops"),
        require("../modules/test_questions")
      ],
      dataSources,
      context
    })
  );
}
