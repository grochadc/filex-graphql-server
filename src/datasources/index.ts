import { FirebaseAPI } from "./myREST";
import { RegistroAPI } from "./RegistroAPI";
import { ExamAPI } from "./ExamAPI";
import { WorkshopsAPI } from "./WorkshopsAPI";
import { StudentsAPI } from "./StudentsAPI";
import { PlacementAPI } from "./PlacementAPI";

export {
  RegistroAPI,
  FirebaseAPI,
  ExamAPI,
  WorkshopsAPI,
  StudentsAPI,
  PlacementAPI,
};
