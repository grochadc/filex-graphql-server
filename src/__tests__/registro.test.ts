import testServer from "../testUtils/testServer";
import Database from "../testUtils/databaseCreator";
import { gql } from "apollo-server";
const { RegistroAPI } = require("../datasources/registroAPI");
import { SheetsAPI } from "../datasources/SheetsAPI";

describe("registro", () => {
  const registroAPI = new RegistroAPI();
  const applicant = {
    codigo: "1234567890",
    nombre: "Benito Antonio",
    apellido_paterno: "Martinez",
    apellido_materno: "Ocasio",
    genero: "M",
    ciclo: "2021A",
    carrera: "Agronegocios",
    telefono: "3411234567",
    email: "bad@bunny.pr",
    externo: true,
    nivel: "4",
    curso: "en"
  };
  const schedulesLevel4 = {
    "E4-1": {
      teacher: "GONZALO ROCHA",
      group: "E4-1",
      sesiones: "HTTP://URL_AQUI",
      chat: "HTTP://URL_AQUI",
      classroom: "HTTP://URL_AQUI"
    }
  };

  it("gets registering levels", async () => {
    const db = {
      prod: {
        registeringLevels: {
          en: ["1", "2"]
        }
      }
    };
    const GET_LEVELS_REGISTERING = gql`
      query getLevelsRegistering($course: String!) {
        english: registeringLevels(course: $course)
      }
    `;
    const variables = { course: "en" };
    const database = new Database(db);
    registroAPI.get = jest.fn(url => database.get(url));
    const dataSources = () => {
      return { registroAPI };
    };
    const context = () => {
      return { enviroment: "prod" };
    };
    const { query } = testServer(dataSources, context);
    const res = await query({
      query: GET_LEVELS_REGISTERING,
      variables
    });
    expect(registroAPI.get).not.toHaveBeenCalledWith(
      expect.stringContaining("undefined")
    );
    expect(registroAPI.get).toHaveBeenCalledTimes(1);
    expect(registroAPI.get).toHaveBeenCalledWith(
      "prod/registeringLevels/en.json"
    );
    expect(res.errors).toBe(undefined);
    expect(res.data).toMatchSnapshot();
  });

  it("gets an applicant who isn't registered", async () => {
    const db = {
      prod: {
        applicants: {
          "1234567890": applicant
        },
        registeringLevels: { en: ["4"] },
        schedules: {
          en: {
            level4: schedulesLevel4
          }
        }
      }
    };
    const GET_APPLICANT = gql`
      query info($codigo: ID!) {
        applicant(codigo: $codigo) {
          codigo
          nombre
          apellido_materno
          apellido_paterno
          genero
          carrera
          ciclo
          telefono
          email
          nivel
          curso
          externo
          registering
          registeredSchedule {
            teacher
            group
            classroom
            chat
            sesiones
          }
          schedules {
            teacher
            group
            serialized(options: { teacher: true, group: true })
          }
        }
      }
    `;
    const variables = {
      codigo: "1234567890"
    };
    const database = new Database(db);
    registroAPI.get = jest.fn(url => database.get(url));
    const dataSources = () => {
      return { registroAPI };
    };
    const context = () => {
      return { enviroment: "prod" };
    };
    const { query } = testServer(dataSources, context);
    const res = await query({
      query: GET_APPLICANT,
      variables
    });
    expect(registroAPI.get).toHaveBeenCalledTimes(5);
    expect(res.errors).toBe(undefined);
    expect(res.data).toMatchSnapshot();
    expect(registroAPI.get).not.toHaveBeenCalledWith(
      expect.stringMatching("undefined")
    );
  });
  it("gets an applicant who is already registered", async () => {
    const db = {
      prod: {
        alreadyRegistered: {
          "1234567890": "E4-1"
        },
        applicants: {
          "1234567890": applicant
        },
        registeringLevels: {
          en: ["3"]
        },
        schedules: {
          en: {
            level4: schedulesLevel4
          }
        }
      }
    };
    const GET_REGISTERED_APPLICANT = gql`
      query info($codigo: ID!) {
        applicant(codigo: $codigo) {
          codigo
          nombre
          email
          nivel
          registering
          registeredSchedule {
            teacher
            group
            classroom
            chat
            sesiones
          }
          schedules {
            teacher
            group
            serialized(options: { teacher: true, group: true })
          }
        }
      }
    `;
    const variables = {
      codigo: "1234567890"
    };

    const database = new Database(db);
    registroAPI.get = jest.fn(url => {
      return database.get(url);
    });
    const dataSources = () => {
      return { registroAPI };
    };
    const context = () => {
      return { enviroment: "prod" };
    };
    const { query } = testServer(dataSources, context);
    const res = await query({
      query: GET_REGISTERED_APPLICANT,
      variables
    });
    expect(registroAPI.get).not.toHaveBeenCalledWith(
      expect.stringContaining("undefined")
    );
    expect(res.data.applicant.registering).toBe(false);
    expect(res.data).toMatchSnapshot();
    expect(res.errors).toBe(undefined);
  });

  it("gets a schedule", async () => {
    const db = {
      prod: {
        schedules: {
          en: {
            level1: {
              "E1-1": {
                teacher: "GONZALO ROCHA",
                group: "E1-1",
                sesiones: "HTTP://URL_AQUI",
                chat: "HTTP://URL_AQUI",
                classroom: "HTTP://URL_AQUI"
              }
            }
          }
        }
      }
    };
    const GET_SCHEDULE = gql`
      query getSchedule($id: String!) {
        schedule(id: $id) {
          group
          teacher
          chat
          classroom
          sesiones
        }
      }
    `;
    const variables = {
      id: "E1-1"
    };
    const database = new Database(db);
    registroAPI.get = jest.fn(url => {
      return database.get(url);
    });
    const dataSources = () => {
      return { registroAPI };
    };
    const context = () => {
      return { enviroment: "prod" };
    };
    const { query } = testServer(dataSources, context);
    const res = await query({
      query: GET_SCHEDULE,
      variables
    });
    expect(registroAPI.get).not.toHaveBeenCalledWith(
      expect.stringContaining("undefined")
    );
    expect(res.data).toMatchSnapshot();
    expect(res.errors).toBe(undefined);
  });

  it("saves levels registering", async () => {
    const SAVE_LEVELS_REGISTERING = gql`
      mutation saveLevels($levels: [String]!, $course: String!) {
        saveRegisteringLevels(levels: $levels, course: $course)
      }
    `;

    const variables = {
      levels: ["1", "2"],
      course: "en"
    };
    registroAPI.put = jest.fn();

    const dataSources = () => {
      return { registroAPI };
    };
    const context = () => {
      return { enviroment: "prod" };
    };
    const { query } = testServer(dataSources, context);
    const res = await query({
      query: SAVE_LEVELS_REGISTERING,
      variables
    });
    expect(registroAPI.put).not.toHaveBeenCalledWith(
      expect.stringContaining("undefined")
    );
    expect(
      registroAPI.put
    ).toHaveBeenCalledWith("prod/registeringLevels/en.json", ["1", "2"]);
    expect(res.data).toMatchSnapshot();
    expect(res.errors).toBe(undefined);
  });

  it("registers a student", async () => {
    const db = {
      prod: {
        schedules: {
          en: {
            level4: schedulesLevel4
          }
        }
      }
    };
    const REGISTER_STUDENT = gql`
      mutation register(
        $codigo: ID!
        $nombre: String!
        $apellido_materno: String!
        $apellido_paterno: String!
        $genero: String!
        $carrera: String!
        $ciclo: String!
        $telefono: String!
        $email: String!
        $nivel: String!
        $curso: String!
        $externo: Boolean!
        $schedule: String!
      ) {
        registerStudent(
          input: {
            codigo: $codigo
            nombre: $nombre
            apellido_materno: $apellido_materno
            apellido_paterno: $apellido_paterno
            genero: $genero
            carrera: $carrera
            ciclo: $ciclo
            telefono: $telefono
            email: $email
            nivel: $nivel
            curso: $curso
            externo: $externo
            grupo: $schedule
          }
        ) {
          nombre
          schedule {
            group
            teacher
            chat
            classroom
            sesiones
          }
        }
      }
    `;
    const variables = {
      codigo: "1234567890",
      nombre: "Benito Antonio",
      apellido_materno: "Martinez",
      apellido_paterno: "Ocasio",
      genero: "M",
      carrera: "Abogado",
      ciclo: "2021A",
      telefono: "1234567890",
      email: "bad@bunny.pr",
      nivel: "4",
      curso: "en",
      externo: false,
      schedule: "E4-1"
    };
    const database = new Database(db);
    const registroAPI = new RegistroAPI();
    registroAPI.get = jest.fn(url => database.get(url));
    registroAPI.post = jest.fn((url, data) => Promise.resolve());

    const registroSheetsAPI = new SheetsAPI("id");
    registroSheetsAPI.append = jest.fn(() => Promise.resolve());

    const dataSources = () => {
      return {
        registroAPI: registroAPI,
        registroSheetsAPI: registroSheetsAPI
      };
    };

    const context = () => {
      return {
        enviroment: "prod"
      };
    };
    const { query } = testServer(dataSources, context);
    const res = await query({
      query: REGISTER_STUDENT,
      variables
    });
    expect(registroAPI.get).not.toHaveBeenCalledWith(
      expect.stringContaining("undefined")
    );
    expect(registroAPI.post).not.toHaveBeenCalledWith(
      expect.stringContaining("undefined")
    );
    expect(registroAPI.post).toHaveBeenCalledWith(
      "prod/availableGroups/en/level4/E4-1.json",
      "1"
    );
    if (res.errors) console.log(JSON.stringify(res.errors));
    expect(res.errors).toBeUndefined();
    expect(res.data).toMatchSnapshot();
  });
});
