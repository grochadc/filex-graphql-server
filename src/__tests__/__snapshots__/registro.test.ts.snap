// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`E2E Gets the levels that are registering 1`] = `
Array [
  1,
  4,
]
`;

exports[`Integration Correctly registers a student 1`] = `
Object {
  "registerStudent": Object {
    "nombre": "Juan",
    "schedule": Object {
      "group": "E4-1",
      "teacher": "Gonzalo Rocha",
      "time": null,
    },
  },
}
`;
